File: src/imagebuilder/flyingtux/services/webproxy.py
from metux.util.fs import mkdir
from os_service import Base
from os.path import basename, dirname

class WebProxy(Base):

    def write_acl(self, fn):
        acl_ip = ('%s_ip' % self.my_appname)
        mkdir(dirname(fn))
        with open(fn, "w") as f:
            f.write("acl %s src %s\n" % (acl_ip, self.my_runner.ip_address))
            cnt = 0
            for wl in self['whitelist']:
                acl_url = '%s_wl%d' % (self.my_appname, cnt)
                f.write("acl %s url_regex %s\n" % (acl_url, wl))
                f.write("http_access allow %s %s\n" % (acl_ip, acl_url))
                cnt = cnt+1

    def compute(self):
        self.my_appname = self.get_app_name()
        proxy = self.my_runner.start_web_proxy()
        self.info("web proxy is: "+proxy)
        self.add_env('http_proxy', proxy)
        self.add_env('https_proxy', proxy)
        self.add_env('ftp_proxy', proxy)
#        self.write_acl(
#            self.my_runner['TARGET::dynconf-dir']+
#            '/squid/%s.conf' % self.my_appname)
        self.write_acl(self.my_runner['TARGET::web-proxy::conf-source']+'/%s.conf' % self.my_appname)

File: internal/services/webproxy.go
// SPDX-License-Identifier: AGPL-3.0-or-later

package services

import (
	"fmt"
	"os"
	"path/filepath"

	"github.com/metux/flyingtux/internal/fs"
	"github.com/metux/flyingtux/internal/spec"
)

// webProxy configures a per-app squid web proxy and its ACL. Port of
// services/webproxy.py.
type webProxy struct {
	base
	appname string
}

func newWebProxy(sp *spec.SpecObj, r Runner) Service {
	s := &webProxy{base: base{name: "WebProxy", spec: sp, runner: r}}
	s.self = s
	s.init()
	return s
}

func (s *webProxy) compute() error {
	s.appname = s.appName()
	proxy, err := s.runner.StartWebProxy()
	if err != nil {
		return err
	}
	s.info("web proxy is: " + proxy)
	s.addEnv("http_proxy", proxy)
	s.addEnv("https_proxy", proxy)
	s.addEnv("ftp_proxy", proxy)
	return s.writeACL(s.runner.Get("TARGET::web-proxy::conf-source") + "/" + s.appname + ".conf")
}

func (s *webProxy) writeACL(fn string) error {
	aclIP := s.appname + "_ip"
	if err := fs.Mkdir(filepath.Dir(fn)); err != nil {
		return err
	}
	f, err := os.Create(fn)
	if err != nil {
		return err
	}
	defer f.Close()

	fmt.Fprintf(f, "acl %s src %s\n", aclIP, s.runner.IPAddress())
	for cnt, wl := range s.spec.GetStrList("whitelist") {
		aclURL := fmt.Sprintf("%s_wl%d", s.appname, cnt)
		fmt.Fprintf(f, "acl %s url_regex %s\n", aclURL, wl)
		fmt.Fprintf(f, "http_access allow %s %s\n", aclIP, aclURL)
	}
	return nil
}
